### PR TITLE
Preserving line endings in blockinfile module

### LINKED ISSUE #64966: Ansible blockinfile ^M bug!?
### SUMMARY
In SuSe SLES there is a file: /etc/vimrc. It contains ^M characters, like in line:  101 containing:
    map! <ESC>OM  ^M
When running module blockinfile on this file it removes the ^M characters.
The blockinfile just needs to adds some lines on the end of the file. Even when the lines are already there it removes the ^M characters.

It looks like that it reads the file and cannot handle the ^M characters and writes the buffer without the characters to the file.

##### ISSUE TYPE
- Module blockinfile

##### COMPONENT NAME
blockinfile

##### ANSIBLE VERSION

```
ansible 2.4.2.0
  config file = None
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /bin/ansible
  python version = 2.7.5 (default, Aug  7 2019, 00:51:29) [GCC 4.8.5 20150623 (Red Hat 4.8.5-39)]


```

##### CONFIGURATION

```paste below

```

##### STEPS TO REPRODUCE

```yaml

```

##### ACTUAL RESULTS

```paste below

```

### PR DESCRIPTION
**Pull request by riyad**
##### SUMMARY
Preserve line endings in blockinfile module.

Fixes #64966

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME
blockinfile

##### ADDITIONAL INFORMATION
I'm not sure if this is the desired way to fix this ... it looks kinda "hacky", but OTOH allows us to remove other hacks (e.g. fixing new lines at the end of files).

In #65677 it was suggested the changed line endings might be related to missing `newline=''` parameter when calling `open()`. But trying to add this you get the following error: `ValueError: binary mode doesn't take a newline argument`. Instead it looks like `splitlines()` is the actual culprit ... at least in this case. :neutral_face:

### CODE REVIEW

**Review by jamescassell — APPROVED** [review]
shipit

**Comment by riyad** [comment]
Updates:
- fixed Python 2.7 compatibility
- fixed BSD/macOS compatibility in tests
- added changelog fragment

The only failure is for the test file with the mixed line endings. :neutral_face: 

Can I mark the test file in any way to skip/ignore that sanity check?
Should I squash the changes?

**Comment by riyad** [comment]
@jamescassell could you have a look at my questions in my last comment? I'd appreciate it. :smiley:

**Comment by jamescassell** [comment]
@riyad don't squash; that happens in merge regardless.

@mattclay how should we handle the mixed line endings? Can we tell that test to skip this file, or should we generate the file at runtime?

**Comment by mattclay** [comment]
@jamescassell Generate the file at runtime.

**Comment by riyad** [comment]
@jamescassell I've updated the PR as requested ... all CI tests pass now. What can I do to get the changes reviewed?

**Comment by riyad** [comment]
Any chance for a review?

**Comment by riyad** [comment]
@jamescassell @mattclay Can you please (re-)review this PR?

**Comment by jamescassell** [comment]
shipit

But this might have moved repos? At least merge for 2.9, then open another PR if necessary elsewhere.

**Comment by riyad** [comment]
I've rebased this PR to make ansibot happy.

**Comment by riyad** [comment]
@jamescassell I've rebased the PR onto `devel` ... could you please give it a final review?

File: changelogs/fragments/66461-blockinfile_preserve_line_endings.yaml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - blockingfile - preserve line endings on update (https://github.com/ansible/ansible/issues/64966)
```
Comment (samdoran) [inline changelogs/fragments/66461-blockinfile_preserve_line_endings.yaml]:
```suggestion
  - blockinfile - preserve line endings on update (https://github.com/ansible/ansible/issues/64966)
```

**Review by samdoran — APPROVED** [review]

File: changelogs/fragments/66461-blockinfile_preserve_line_endings.yaml (hunk 1 above)
Comment (riyad) [reply changelogs/fragments/66461-blockinfile_preserve_line_endings.yaml]: embarrassing :sweat_smile:

Comment (samdoran) [reply changelogs/fragments/66461-blockinfile_preserve_line_endings.yaml]: I do stuff like that all the time. 😊

**Comment by riyad** [comment]
Thank you :slightly_smiling_face:

**Comment by samdoran** [comment]
This is causing test failures, so this has introduced a bug. This task:
https://github.com/ansible/ansible/blob/e5cc12a64f274e36738fe51d4caa533003ed626b/test/integration/targets/prepare_http_tests/tasks/default.yml#L45-L49

is producing the following incorrect output:

(this is truncated)
```
...
LWQA9Lzg23bzXcA2kkImYau9Ee9qEt7Cw3mNUttwGFispNKd+UDQ5QHT00jqRYBs
cE3qUFqvqUqTCJAxbIjO8YD56IHw4iPLo4m6k7H8wuug1k/VEhGe6WhZa1s2ZVgV
PYoE9J+1ega9Q3ZdrT4+eT0XJg==
-----END CERTIFICATE-----# END ANSIBLE MANAGED BLOCK
# BEGIN ANSIBLE MANAGED BLOCK
-----BEGIN CERTIFICATE-----
MIIDnzCCAoegAwIBAgIJAOTtpfAnv5rGMA0GCSqGSIb3DQEBCwUAMGYxCzAJBgNV
BAYTAlVTMRcwFQYDVQQIDA5Ob3J0aCBDYXJvbGluYTEPMA0GA1UEBwwGRHVyaGFt
MRAwDgYDVQQKDAdBbnNpYmxlMRswGQYDVQQDDBJhbnNpYmxlLmh0dHAu
...
```

Looks like we aren't preserving the _leading_ line ending.

I'll see if I can come up with a quick fix plus tests, otherwise we will need to revert this.

**Comment by samdoran** [comment]
I have a fix. I'll have a PR shortly. Writing tests...

File: lib/ansible/modules/blockinfile.py [hunk 2]
```diff
@@ -306,9 +306,7 @@ def main():
     lines[n0:n0] = blocklines
 
     if lines:
-        result = b('\n').join(lines)
```
Comment (samdoran) [inline lib/ansible/modules/blockinfile.py:309]: This code ensured that the block text always had a line separator after it. With this removed, a block that has no line separator ends up being concatenated to the end mark.

Comment (samdoran) [reply lib/ansible/modules/blockinfile.py:309]: Sorry I missed that in review.

**Pull request merged**